File: aac/api/typescript/src/protocol/MessageParser.ts
/**
 * WIA AAC Protocol Message Parser
 */

import {
  WiaAacMessage,
  MessageType,
  PROTOCOL_NAME,
  PROTOCOL_VERSION,
  ErrorCodes,
  ProtocolError,
} from './message';
import { MessageBuilder } from './MessageBuilder';

export interface ParseResult<T = unknown> {
  success: boolean;
  message?: WiaAacMessage<T>;
  error?: ProtocolError;
}

export class MessageParser {
  private supportedVersions: string[];

  constructor(supportedVersions: string[] = [PROTOCOL_VERSION]) {
    this.supportedVersions = supportedVersions;
  }

  /**
   * Parse a JSON string into a WiaAacMessage
   */
  parse(data: string | object): ParseResult {
    try {
      // Parse JSON if string
      const obj = typeof data === 'string' ? JSON.parse(data) : data;

      // Validate required fields
      const validationError = this.validate(obj);
      if (validationError) {
        return {
          success: false,
          error: validationError,
        };
      }

      return {
        success: true,
        message: obj as WiaAacMessage,
      };
    } catch (e) {
      return {
        success: false,
        error: MessageBuilder.createError(
          ErrorCodes.INVALID_MESSAGE,
          'INVALID_MESSAGE',
          `Failed to parse message: ${e instanceof Error ? e.message : 'Unknown error'}`,
          false
        ),
      };
    }
  }

  /**
   * Validate a message object
   */
  private validate(obj: unknown): ProtocolError | null {
    if (!obj || typeof obj !== 'object') {
      return MessageBuilder.createError(
        ErrorCodes.INVALID_MESSAGE,
        'INVALID_MESSAGE',
        'Message must be an object',
        false
      );
    }

    const msg = obj as Record<string, unknown>;

    // Check protocol
    if (msg.protocol !== PROTOCOL_NAME) {
      return MessageBuilder.createError(
        ErrorCodes.PROTOCOL_ERROR,
        'PROTOCOL_ERROR',
        `Invalid protocol: expected '${PROTOCOL_NAME}', got '${msg.protocol}'`,
        false
      );
    }

    // Check version
    if (typeof msg.version !== 'string') {
      return MessageBuilder.createError(
        ErrorCodes.INVALID_MESSAGE,
        'INVALID_MESSAGE',
        'Missing or invalid version field',
        false
      );
    }

    if (!this.isVersionSupported(msg.version)) {
      return MessageBuilder.createError(
        ErrorCodes.VERSION_MISMATCH,
        'VERSION_MISMATCH',
        `Unsupported protocol version: ${msg.version}`,
        true,
        { supportedVersions: this.supportedVersions }
      );
    }

    // Check messageId
    if (typeof msg.messageId !== 'string' || !msg.messageId) {
      return MessageBuilder.createError(
        ErrorCodes.INVALID_MESSAGE,
        'INVALID_MESSAGE',
        'Missing or invalid messageId field',
        false
      );
    }

    // Check timestamp
    if (typeof msg.timestamp !== 'number' || msg.timestamp < 0) {
      return MessageBuilder.createError(
        ErrorCodes.INVALID_MESSAGE,
        'INVALID_MESSAGE',
        'Missing or invalid timestamp field',
        false
      );
    }

    // Check type
    if (!this.isValidMessageType(msg.type)) {
      return MessageBuilder.createError(
        ErrorCodes.INVALID_MESSAGE,
        'INVALID_MESSAGE',
        `Invalid message type: ${msg.type}`,
        false
      );
    }

    // Check payload
    if (typeof msg.payload !== 'object') {
      return MessageBuilder.createError(
        ErrorCodes.INVALID_MESSAGE,
        'INVALID_MESSAGE',
        'Missing or invalid payload field',
        false
      );
    }

    return null;
  }

  /**
   * Check if version is supported
   */
  private isVersionSupported(version: string): boolean {
    // Check exact match
    if (this.supportedVersions.includes(version)) {
      return true;
    }

    // Check major version compatibility (1.x.x is compatible with 1.0.0)
    const [major] = version.split('.');
    return this.supportedVersions.some((v) => v.startsWith(`${major}.`));
  }

  /**
   * Check if message type is valid
   */
  private isValidMessageType(type: unknown): type is MessageType {
    const validTypes: MessageType[] = [
      'connect',
      'connect_ack',
      'disconnect',
      'disconnect_ack',
      'signal',
      'command',
      'command_ack',
      'subscribe',
      'subscribe_ack',
      'unsubscribe',
      'unsubscribe_ack',
      'error',
      'ping',
      'pong',
    ];
    return typeof type === 'string' && validTypes.includes(type as MessageType);
  }

  /**
   * Serialize a message to JSON string
   */
  serialize(message: WiaAacMessage): string {
    return JSON.stringify(message);
  }
}

// Export singleton instance
export const messageParser = new MessageParser();

File: aac/api/python/wia_aac/protocol/message_parser.py
"""
WIA AAC Protocol Message Parser
"""

import json
from dataclasses import dataclass
from typing import Optional, List, Union, Dict, Any

from .message import (
    PROTOCOL_NAME,
    PROTOCOL_VERSION,
    WiaAacMessage,
    ProtocolError,
    ProtocolErrorCode,
)
from .message_builder import MessageBuilder

VALID_MESSAGE_TYPES = [
    "connect",
    "connect_ack",
    "disconnect",
    "disconnect_ack",
    "signal",
    "command",
    "command_ack",
    "subscribe",
    "subscribe_ack",
    "unsubscribe",
    "unsubscribe_ack",
    "error",
    "ping",
    "pong",
]


@dataclass
class ParseResult:
    """Result of parsing a message"""
    success: bool
    message: Optional[WiaAacMessage] = None
    error: Optional[ProtocolError] = None


class MessageParser:
    """Message parser for WIA AAC protocol"""

    def __init__(self, supported_versions: Optional[List[str]] = None):
        self.supported_versions = supported_versions or [PROTOCOL_VERSION]

    def parse(self, data: Union[str, Dict[str, Any]]) -> ParseResult:
        """Parse a JSON string or dict into a WiaAacMessage"""
        try:
            # Parse JSON if string
            if isinstance(data, str):
                obj = json.loads(data)
            else:
                obj = data

            # Validate
            validation_error = self._validate(obj)
            if validation_error:
                return ParseResult(success=False, error=validation_error)

            # Create message
            message = WiaAacMessage.from_dict(obj)
            return ParseResult(success=True, message=message)

        except json.JSONDecodeError as e:
            return ParseResult(
                success=False,
                error=MessageBuilder.create_error(
                    ProtocolErrorCode.INVALID_MESSAGE,
                    "INVALID_MESSAGE",
                    f"Failed to parse JSON: {e}",
                    False,
                ),
            )
        except Exception as e:
            return ParseResult(
                success=False,
                error=MessageBuilder.create_error(
                    ProtocolErrorCode.INVALID_MESSAGE,
                    "INVALID_MESSAGE",
                    f"Failed to parse message: {e}",
                    False,
                ),
            )

    def _validate(self, obj: Any) -> Optional[ProtocolError]:
        """Validate a message object"""
        if not obj or not isinstance(obj, dict):
            return MessageBuilder.create_error(
                ProtocolErrorCode.INVALID_MESSAGE,
                "INVALID_MESSAGE",
                "Message must be an object",
                False,
            )

        # Check protocol
        if obj.get("protocol") != PROTOCOL_NAME:
            return MessageBuilder.create_error(
                ProtocolErrorCode.PROTOCOL_ERROR,
                "PROTOCOL_ERROR",
                f"Invalid protocol: expected '{PROTOCOL_NAME}', got '{obj.get('protocol')}'",
                False,
            )

        # Check version
        version = obj.get("version")
        if not isinstance(version, str):
            return MessageBuilder.create_error(
                ProtocolErrorCode.INVALID_MESSAGE,
                "INVALID_MESSAGE",
                "Missing or invalid version field",
                False,
            )

        if not self._is_version_supported(version):
            return MessageBuilder.create_error(
                ProtocolErrorCode.VERSION_MISMATCH,
                "VERSION_MISMATCH",
                f"Unsupported protocol version: {version}",
                True,
                {"supportedVersions": self.supported_versions},
            )

        # Check messageId
        if not obj.get("messageId") or not isinstance(obj.get("messageId"), str):
            return MessageBuilder.create_error(
                ProtocolErrorCode.INVALID_MESSAGE,
                "INVALID_MESSAGE",
                "Missing or invalid messageId field",
                False,
            )

        # Check timestamp
        timestamp = obj.get("timestamp")
        if not isinstance(timestamp, (int, float)) or timestamp < 0:
            return MessageBuilder.create_error(
                ProtocolErrorCode.INVALID_MESSAGE,
                "INVALID_MESSAGE",
                "Missing or invalid timestamp field",
                False,
            )

        # Check type
        msg_type = obj.get("type")
        if msg_type not in VALID_MESSAGE_TYPES:
            return MessageBuilder.create_error(
                ProtocolErrorCode.INVALID_MESSAGE,
                "INVALID_MESSAGE",
                f"Invalid message type: {msg_type}",
                False,
            )

        # Check payload
        if not isinstance(obj.get("payload"), dict):
            return MessageBuilder.create_error(
                ProtocolErrorCode.INVALID_MESSAGE,
                "INVALID_MESSAGE",
                "Missing or invalid payload field",
                False,
            )

        return None

    def _is_version_supported(self, version: str) -> bool:
        """Check if version is supported"""
        # Check exact match
        if version in self.supported_versions:
            return True

        # Check major version compatibility
        major = version.split(".")[0]
        return any(v.startswith(f"{major}.") for v in self.supported_versions)

    def serialize(self, message: WiaAacMessage) -> str:
        """Serialize a message to JSON string"""
        return json.dumps(message.to_dict())


# Singleton instance
message_parser = MessageParser()
